specs/marketplace.md

```diff
@@ -0,0 +1,381 @@
+extend the Request's duration, somebody submits a new Request transaction with the same CID **well before the original
+Request finishes**. In this way, the data will be still persisted in the network at the time when new (or the current) storage providers 
+need to retrieve the dataset to fill slots of the new Request.
+
+### Withdrawing funds
+
+The client node SHOULD monitor the status of Requests that it created. The node can utilize on-chain state to
+fetch the list of the active Requests linked to the client node's ethereum address using function `myRequests()`, that
+returns array of `RequestId`s. This list is kept up to date by the smart contract itself.
+
+When Request reaches states `Cancelled` (not all slots filled after `expiry` timeout) or `Failed` (too many slots gets freed and data is non-recoverable)
+the client node SHOULD initiate withdrawal of the remaining funds from the contract using function `withdrawFunds(requestId)`.
+
+ - `Cancelled` state MAY be detected using timeout specified from function `requestExpiresAt(requestId)` **and** not detecting emitted `RequestFulfilled(requestId)` event.
+ - `Failed` state MAY be detected using `RequestFailed(requestId)` event emitted from the smart contract.
+ - `Finished` state MAY be detected by setting timeout specified from function `getRequestEnd(requestId)`.
+
+## Storage Provider role
+
+Storage Provider (SP) role represents nodes that persist data across the network by hosting Slots of Requests 
+that Client nodes requested.
+
+There are several parts to hosting a slot:
+
+ - Filling a slot
+ - Proving
+ - Repairing a slot
+ - Collecting Request's reward and collateral
+
+### Filling slot
+
+When new Request is created `StorageRequested(requestId, ask, expiry)` event is emitted with following properties: 
+
+ - `requestId` - ID of Request.
+ - `ask` - Specification of Request parameters. For details see above.
+ - `expiry` - Unix timestamp that specifies when the Request will be cancelled if all slots are not filled by then.
+
+It is then up to the Storage Provider node to decide based on the emitted parameters if it wants to participate in the 
+Request and try to fill its slot(s). This decision SHOULD be done based on parameters specified by the node operator.
+If the node decides to ignore this Request, no action is necessary. If the node wants to try to fill a a slot, then 
+it MUST follow the remaining steps.
+
+Node MUST decide which Slot specified by slot's index it wants to try to fill in. Node MAY try filling multiple
+slots. In order to fill a slot, node first MUST download the slot's data using slot's root that can be retrieved
+from Manifest specified in `request.content.cid` (**TODO: Manifest RFC**). 
+This Request object can be retrieved from the smart contract using `getRequest(requestId)`.
+Then the node MUST generate proof over the downloaded data (**TODO: Proving RFC**).
```

Comment:
It also needs to get the challenge and merkleRoot from the contract as inputs to the prover.